Lab3 - Add Letter frequency and sort letters functions

davidbroughsmyth opened the pull request:
Fixes #544 

Review of colleague's PR #339 

#### Changes proposed in this PR:
- Concurrent implementation Letter frequency `letters(s string) map[rune]int`
- Standard Letter frequency implementation `letterFreq(s string) map[rune]int`
- Function that returns a sorted slice of strings with elements "{key}:{val}"
- Provide comprehensive tests and benchmark for standard and Concurrent Letter frequency functions

davidbroughsmyth committed:
Complete Lab3 Letter frequency

Provide tests and benchmarks
BenchmarkLetterFreq-4                 50000         24113 ns/op
BenchmarkConcurrentLetters-4          20000         90724 ns/op

camscale commented on 03_letters/davidbroughsmyth/main.go:
```diff
@@ -0,0 +1,56 @@
+	return f
+}
+
+// Concurrent implementation of letters for speed
```
Go doc strings have a particular format that should be followed: https://blog.golang.org/godoc-documenting-go-code.

The comment "Concurrent implementation of letters for speed" does not belong here either as that would end up in the documentation which is not relevant to callers of the function.

davidbroughsmyth replied:
I've removed this comment

camscale commented on 03_letters/davidbroughsmyth/main.go:
```diff
@@ -0,0 +1,56 @@
+		}(ss)
+	}
+
+	//Combine the letter frequency maps passed back via the results channel
```
nitpick: You're using mixed styles for comments - some with `// ` (slash-slash-space) and some with `//` (slash-slash). I suggest you stick to the former as it is the most common Go style.

davidbroughsmyth replied:
Fixed this, it was unintentional and slipped through.

camscale commented on 03_letters/davidbroughsmyth/main.go:
```diff
@@ -0,0 +1,56 @@
+
+	//Combine the letter frequency maps passed back via the results channel
+	for i := 0; i < size; i++ {
+		for r, freq := range <-results {
```
This is a little confusing because it kind of conflates ranging over a map and ranging over a channel. I think it would be clearer to add a line:
```
result := <-results
for r, freq := range result {
```
Go has deliberately omitted certain features (ternary operator, post inc/dec in expressions, etc) as they often lead to confusing code, so good Go code generally aims to keep it simple. Splitting the channel read and the map iteration keeps it simple. And since you have benchmarks (cool), you can see it makes no difference to this generated code.

davidbroughsmyth replied:
I like this better too. The change has been made.

camscale commented on 03_letters/davidbroughsmyth/main_test.go:
```diff
@@ -0,0 +1,126 @@
+		wantLetters: map[rune]int{'♞': 1, '🁻': 1, '∑': 1, '👾': 2, '😸': 2, '😛': 3},
+		wantSort:    []string{"∑:1", "♞:1", "🁻:1", "👾:2", "😛:3", "😸:2"},
+	},
+}
```
You have no tests that have spaces in the input, yet the code uses `String.Fields` to split on spaces. So none of the tests are using much in the way of concurrency and that is the more complicated part of your code. You really want to ensure that is being tested more rigorously.

davidbroughsmyth replied:
I've added specific tests with spaces.

camscale commented on 03_letters/davidbroughsmyth/main_test.go:
```diff
@@ -0,0 +1,126 @@
+
+func TestLetterFreq(t *testing.T) {
+	for name, test := range tests {
+		testList := test
```
`testList` is a funny name for the test case. If you were struggling to find an appropriate name, you can just re-use `test` - this is pretty common in this case:
```
test := test
```
It initially looks a bit weird, but that's useful because it highlights a weird aspect of Go closures over loop variables.

davidbroughsmyth replied:
I've made this change. It does look weird, but I also like the way closures work in Go.

camscale commented on 03_letters/davidbroughsmyth/main.go:
```diff
@@ -0,0 +1,56 @@
+// Returns the letter frequency for string
+func letters(s string) map[rune]int {
+	stringSlice := strings.Fields(s)
+	combineFreq := map[rune]int{}
```
I'd suggest moving this closer to where it is used (just above the loop that reads the results). This looks too much like C-style - declaring vars at the top of the function, and is not idiomatic go - keeping the "range" of a variable short so you can stick to shorter names is the usual practice.

davidbroughsmyth replied:
I've done this change.

camscale reviewed (dismissed):
A couple of Go-oriented comments, plus some other general software engineering oriented ones. I let the latter ones slide, but not the former. Since you have some of the former, you may as well address them all.

However since you have gone with a more complicated implementation "for speed", were this a real project I'd be asking you to include the results of your benchmarking in the git commit message. More complicated code adds a maintenance burden and it needs to be justified. Optimising without showing the need for it should be avoided.

Even better would be two commits - the first with the basic implementation and the second with the faster one, with the commit message having the benchmark details. That way you have a baseline for the comparison. The tests in the first commit would be comprehensive enough that they apply unchanged to the concurrent implementation. But I won't hold you to that - it's just an idea and how I'd probably structure my commits in this case.

camscale commented:
Could you also update your commit message to adhere to the "Recipe for great commit message" as outlined in the course slides: https://go-course.org/gocourse-2019-05-melbourne.slide#31

I let it slide on lab 2, but since you need to address comments on this commit, it's time to get that right. Thanks.

davidbroughsmyth committed:
Fix changes requested in PR

Removed unnecessary comments
Fix mixed styles for comments
Split the channel read from the map iteration
Moved sum definition closer to where its used
Provide tests with spaces
Use test case name more suitable for Go style

davidbroughsmyth commented:
I've updated the commit message and add benchmark data showing the speed improvement.

davidbroughsmyth commented:
The speed I expected when implementing this Concurrent version of Letter frequency was not achieved, the benchmarks show that it is about 4 times slower than the standard synchronous version. Which is fine by me. I learned so much more about the go language and goroutines than any of the labs so far. I would like to keep the implementation as-is, assuming it meets the code owners criteria.

juliaogris reviewed (approved):
Looks great!

Mini-nit: I've updated your PR description Markdown (empty line after bulletin adds undesired paragraph)

davidbroughsmyth committed:
Merge branch 'master' into lab3
